caffe2/python/operator_test/elementwise_ops_test.py

```diff
@@ -376,49 +378,86 @@ def ref(A, B):
 
         A = np.random.rand(n, m, 1, 1).astype(np.float32) + bias
         B = np.random.rand(n, m, k, t).astype(np.float32) + bias
-        self._run_single_test(op, ref, A, B, test_grad, gc, dc)
+        self._run_single_test(op, ref, A, B, True, test_grad, gc, dc)
 
         A = np.random.rand(m, 1, t).astype(np.float32) + bias
         B = np.random.rand(n, m, k, t).astype(np.float32) + bias
-        self._run_single_test(op, ref, A, B, test_grad, gc, dc)
+        self._run_single_test(op, ref, A, B, True, test_grad, gc, dc)
 
         A = np.random.rand(1, m, 1, t).astype(np.float32) + bias
         B = np.random.rand(n, 1, k, 1).astype(np.float32) + bias
-        self._run_single_test(op, ref, A, B, test_grad, gc, dc)
+        self._run_single_test(op, ref, A, B, True, test_grad, gc, dc)
+
+    def _test_binary_op_in_place(
+            self, op_name, np_ref, n, m, bias, test_grad, in_place_2nd, gc, dc):
+        def ref(A, B):
+            return [np_ref(A, B)]
+
+        op = core.CreateOperator(
+            op_name,
+            ["A", "B"],
+            ["A"],
+        )
+        A = np.random.rand(n, m).astype(np.float32) + bias
+        B = np.random.rand(m).astype(np.float32) + bias
+
+        self._run_single_test(op, ref, A, B, False, test_grad, gc, dc)
+        A = np.random.rand(n, m).astype(np.float32) + bias
+        B = np.random.rand(n, 1).astype(np.float32) + bias
+        self._run_single_test(op, ref, A, B, False, test_grad, gc, dc)
+
+        if in_place_2nd:
+            op = core.CreateOperator(
+                op_name,
+                ["A", "B"],
+                ["B"],
+            )
+            A = np.random.rand(m).astype(np.float32) + bias
+            B = np.random.rand(n, m).astype(np.float32) + bias
+            self._run_single_test(op, ref, A, B, False, test_grad, gc, dc)
+            A = np.random.rand(n, 1).astype(np.float32) + bias
+            B = np.random.rand(n, m).astype(np.float32) + bias
+            self._run_single_test(op, ref, A, B, False, test_grad, gc, dc)
 
     @given(n=st.integers(0, 5), m=st.integers(0, 5), k=st.integers(0, 5),
            t=st.integers(0, 5), **hu.gcs)
     def test_add(self, n, m, k, t, gc, dc):
         self._test_binary_op("Add", np.add, n, m, k, t, -0.5, True, gc, dc)
+        self._test_binary_op_in_place(
+            "Add", np.add, n, m, -0.5, True, True, gc, dc)
 
     @given(n=st.integers(0, 5), m=st.integers(0, 5), k=st.integers(0, 5),
            t=st.integers(0, 5), **hu.gcs)
     def test_sub(self, n, m, k, t, gc, dc):
         self._test_binary_op("Sub", np.subtract, n, m,
                              k, t, -0.5, True, gc, dc)
+        self._test_binary_op_in_place(
+            "Sub", np.subtract, n, m,  -0.5, True, True, gc, dc)
 
     @given(n=st.integers(0, 5), m=st.integers(0, 5), k=st.integers(0, 5),
-           t=st.integers(0, 5), **hu.gcs)
+           t=st.integers(0, 5), **hu.gcs_cpu_only)
```

Comment:
Something is wrong with GPU's `Mul` implementation?

Reply:
Done